# Part-3 Data IO with SeDF - Exporting Data

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></span><ul class="toc-item"><li><span><a href="#Create-a-SeDF" data-toc-modified-id="Create-a-SeDF-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Create a SeDF</a></span></li></ul></li><li><span><a href="#Writing-GIS-Data" data-toc-modified-id="Writing-GIS-Data-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Writing GIS Data</a></span><ul class="toc-item"><li><span><a href="#Publish-as-a-Feature-Layer" data-toc-modified-id="Publish-as-a-Feature-Layer-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Publish as a Feature Layer</a></span></li><li><span><a href="#Write-to-JSON-based-formats" data-toc-modified-id="Write-to-JSON-based-formats-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Write to JSON based formats</a></span><ul class="toc-item"><li><span><a href="#Write-to-FeatureSet" data-toc-modified-id="Write-to-FeatureSet-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Write to FeatureSet</a></span></li><li><span><a href="#Write-to-FeatureCollection" data-toc-modified-id="Write-to-FeatureCollection-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Write to FeatureCollection</a></span></li></ul></li><li><span><a href="#Write-to-a-local-file" data-toc-modified-id="Write-to-a-local-file-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Write to a local file</a></span><ul class="toc-item"><li><span><a href="#Write-to-local-databases" data-toc-modified-id="Write-to-local-databases-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Write to local databases</a></span><ul class="toc-item"><li><span><a href="#File-Geodatabase" data-toc-modified-id="File-Geodatabase-2.3.1.1"><span class="toc-item-num">2.3.1.1&nbsp;&nbsp;</span>File Geodatabase</a></span></li><li><span><a href="#Mobile-Geodatabase" data-toc-modified-id="Mobile-Geodatabase-2.3.1.2"><span class="toc-item-num">2.3.1.2&nbsp;&nbsp;</span>Mobile Geodatabase</a></span></li><li><span><a href="#SQLite-Database" data-toc-modified-id="SQLite-Database-2.3.1.3"><span class="toc-item-num">2.3.1.3&nbsp;&nbsp;</span>SQLite Database</a></span></li></ul></li><li><span><a href="#Write-to-a-shapefile" data-toc-modified-id="Write-to-a-shapefile-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Write to a shapefile</a></span></li><li><span><a href="#Write-to-Non-spatial-formats" data-toc-modified-id="Write-to-Non-spatial-formats-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Write to Non-spatial formats</a></span><ul class="toc-item"><li><span><a href="#Write-to-a-csv-file" data-toc-modified-id="Write-to-a-csv-file-2.3.3.1"><span class="toc-item-num">2.3.3.1&nbsp;&nbsp;</span>Write to a csv file</a></span></li><li><span><a href="#Write-to-a-table-in-a-File-Geodatabase" data-toc-modified-id="Write-to-a-table-in-a-File-Geodatabase-2.3.3.2"><span class="toc-item-num">2.3.3.2&nbsp;&nbsp;</span>Write to a table in a File Geodatabase</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Overwriting-GIS-Data" data-toc-modified-id="Overwriting-GIS-Data-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Overwriting GIS Data</a></span><ul class="toc-item"><li><span><a href="#Overwrite-a-Featureclass" data-toc-modified-id="Overwrite-a-Featureclass-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Overwrite a Featureclass</a></span></li><li><span><a href="#Overwrite-a-table" data-toc-modified-id="Overwrite-a-table-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Overwrite a table</a></span><ul class="toc-item"><li><span><a href="#Table-in-a-csv-file" data-toc-modified-id="Table-in-a-csv-file-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Table in a csv file</a></span></li><li><span><a href="#Table-in-a-File-Geodatabase" data-toc-modified-id="Table-in-a-File-Geodatabase-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Table in a File Geodatabase</a></span></li></ul></li></ul></li><li><span><a href="#Memory-based-Workspace" data-toc-modified-id="Memory-based-Workspace-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Memory-based Workspace</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></div>

## Introduction

In _part-2_ of this guide series, we saw how GIS data can be accessed from various data formats using [__Spatially enabled DataFrame__](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html#geoaccessor) (SeDF). In this part of the guide series, we will look at how SeDF can be used to export the data to various spatial and non-spatial formats. We will also explore how local data can be easily overwritten using SeDF. Let's explore some of the different options available with the versatile _Spatially enabled DataFrame_.

The data used in this guide is provided as an [item](https://www.arcgis.com/home/item.html?id=518fcbd5ef7440a0877726b0ac95b421). We will start by importing some libaries and downloading and extracting the data needed for the analysis in this guide. 

In [1]:
# Import Libraries
import pandas as pd
from arcgis.features import GeoAccessor, GeoSeriesAccessor
from arcgis.gis import GIS
from IPython.display import display
import zipfile
import os
import shutil

In [2]:
# Create a GIS connection
gis = GIS()
agol_gis = GIS("https://www.arcgis.com","arcgis_python","amazing_arcgis_123")

In [6]:
# Get the data item
data_item = gis.content.get('c7140ae3d7ae4fd0817181461019aa75')
data_item

<Item title:"sedf_guide_data" type:Shapefile owner:api_data_owner>

The cell below downloads and extracts the data from the data item to your machine.

In [7]:
# Download and extract the data
def unzip_data():
    """
    This function:
    - creates a directory `sedf_data` to download the data from the item
    - downloads the item as `sedf_guide_data.zip` file in the sedf_data directory
    - unzips and extracts the data to '.\sedf_data\cities'.
    """
    try:
        
        data_dir = os.path.join(os.getcwd(), 'sedf_data')    # path to downloaded data folder
        
        # remove existing cities directory if exists
        if os.path.isdir(data_dir):
            shutil.rmtree(data_dir)
            print(f'Removed existing data directory')
        else:
            os.makedirs(data_dir)
            
        data_item.download(data_dir)    # download the data item
        zipped_file_path = os.path.join(data_dir, 'sedf_guide_data.zip')    # path to zipped file inside data folder

        # unzip the data
        zip_ref = zipfile.ZipFile(zipped_file_path, 'r')
        zip_ref.extractall(data_dir)
        zip_ref.close()
        
        cities_dir = os.path.join(data_dir, 'cities')    # path to new cities directory
        print(f'Dataset unzipped at: {os.path.relpath(cities_dir)}')
        
    except Exception as e:
        print(f'Error unzipping file: {e}')
        

# Extract data
unzip_data()

Dataset unzipped at: sedf_data\cities


### Create a SeDF
Here, we will create a SeDF and then export the data to various data formats.

In [8]:
# Retrieve an item from ArcGIS Online using Item ID value
gis = GIS()
item = gis.content.get("85d0ca4ea1ca4b9abf0c51b9bd34de2e")
item

<Item title:"USA Major Cities" type:Feature Layer Collection owner:esri_dm>

In [9]:
# Obtain the first feature layer from the item
flayer = item.layers[0]

# Use the `from_layer` static method in the 'spatial' namespace on the Pandas' DataFrame
sdf = pd.DataFrame.spatial.from_layer(flayer)

# Check shape
sdf.shape

(3886, 50)

In [10]:
# Check first few records
sdf.head()

AGE_10_14  AGE_15_19  AGE_20_24  AGE_25_34  AGE_35_44  AGE_45_54  \
0       1313       1058        734       2031       1767       1446   
1        890        817        818       1799       1235       1330   
2      12750      13959      16966      32135      27048      29595   
3        790        768        699       1445       1136       1134   
4       3803       3779       3687       7571       5559       4744   

   AGE_55_64  AGE_5_9  AGE_65_74  AGE_75_84  ...  PLACEFIPS  POP2010  \
0       1136     1503        665        486  ...    1601990    13816   
1       1143     1099        721        579  ...    1607840    11899   
2      24177    12933      12176       7087  ...    1608830   205671   
3        935      959        679        464  ...    1611260    10345   
4       3624     4397       2296       1222  ...    1612250    46237   

   POPULATION  POP_CLASS  RENTER_OCC  \
0       15181          6        1271   
1       11946          6        1441   
2      225405          8       33359   
3       10727          6        1461   
4       53942          7        5196   

                                               SHAPE  ST STFIPS VACANT   WHITE  
0  {"x": -12462673.723706163, "y": 5384674.994080...  ID     16    271   13002  
1  {"x": -12506251.313993266, "y": 5341537.793529...  ID     16    318    9893  
2  {"x": -12938676.6836459, "y": 5403597.04949123...  ID     16   6996  182991  
3  {"x": -12667411.402393516, "y": 5241722.820606...  ID     16    241    7984  
4  {"x": -12989383.674504517, "y": 5413226.487333...  ID     16   1428   35856  

[5 rows x 50 columns]

In [11]:
# Check type of sdf
type(sdf)

pandas.core.frame.DataFrame

In [12]:
# Access spatial namespace
sdf.spatial.geometry_type

['point']

> We can see that the dataset has 3886 records and 50 columns. Inspecting the `type` of `sdf` object and accessing the `spatial` namespace shows us that a _Spatially enabled DataFrame_ has been created from all the data in the layer.

## Writing GIS Data

The [__Spatially enabled DataFrame__](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html#geoaccessor) can export data to various data formats for use in other applications. Let's dive into the details of exporting GIS data to various sources.

### Publish as a Feature Layer

Data in a Spatially enabled DataFrame can be exported to [`Feature layers`](https://doc.arcgis.com/en/arcgis-online/share-maps/hosted-web-layers.htm) hosted on [**ArcGIS Online**](https://www.arcgis.com) or [**ArcGIS Enterprise**](http://enterprise.arcgis.com/en/) using the [`to_featurelayer()`](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=to_featurelayer#arcgis.features.GeoAccessor.to_featurelayer) method.

Let's export the `sdf` DataFrame, created above, to a feature layer stored in an ArcGIS Online organization.

In [13]:
# Export to feature layer
lyr = sdf.spatial.to_featurelayer('census_cities_export', gis=agol_gis)
lyr

<Item title:"census_cities_export" type:Feature Layer Collection owner:arcgis_python>

In [14]:
# Check type
type(lyr.layers[0])

arcgis.features.layer.FeatureLayer

> The census_cities_export feature layer has been created at the ArcGIS Online connection specified.

### Write to JSON based formats


Data in a Spatially enabled DataFrame can be exported to JSON based formats, such as [FeatureSet](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=query#arcgis.features.FeatureSet) or [FeatureCollection](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html#featurecollection), using the [`to_featureset()`](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=to_featurelayer#arcgis.features.GeoAccessor.to_featureset) and [`to_feature_collection()`](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=to_featurelayer#arcgis.features.GeoAccessor.to_feature_collection) methods. Let's take a look.

#### Write to FeatureSet

The [`to_featureset()`](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=to_featurelayer#arcgis.features.GeoAccessor.to_featureset) method can be used to export data from a _SeDF_ into a _FeatureSet_.

In [15]:
# Write to FeatureSet
fset_exp = sdf.spatial.to_featureset()

In [16]:
# Check type
type(fset_exp)

arcgis.features.feature.FeatureSet

> A FeatureSet object has been created from the data in the SeDF.

#### Write to FeatureCollection

The [`to_feature_collection()`](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=to_featurelayer#arcgis.features.GeoAccessor.to_feature_collection) method can be used to export data from a _SeDF_ into a _FeatureCollection_.

In [17]:
# Write to FeatureCollection
fc_exp = sdf.spatial.to_feature_collection()

In [18]:
# Check type
type(fc_exp)

arcgis.features.feature.FeatureCollection

> A FeatureCollection object has been created from the data in the SeDF.

### Write to a local file

Data in a Spatially enabled DataFrame can be exported to local spatial file formats, such as [Feature classes](http://desktop.arcgis.com/en/arcmap/latest/manage-data/feature-classes/a-quick-tour-of-feature-classes.htm) or [shapefiles](http://desktop.arcgis.com/en/arcmap/latest/manage-data/shapefiles/what-is-a-shapefile.htm), and non-spatial formats, such as [csv](https://en.wikipedia.org/wiki/Comma-separated_values) files or [tables](https://pro.arcgis.com/en/pro-app/latest/help/data/geodatabases/overview/table-basics.htm). Let's take a look.

#### Write to local databases

The [`to_featureclass()`](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=from_feather#arcgis.features.GeoAccessor.to_featureclass) method can be used to export spatial data from a _SeDF_ into various local databases, such as a [File geodatabase](https://pro.arcgis.com/en/pro-app/latest/help/data/geodatabases/manage-file-gdb/file-geodatabases.htm), a [Mobile geodatabase](https://pro.arcgis.com/en/pro-app/latest/help/data/geodatabases/manage-mobile-gdb/mobile-geodatabases.htm) (.geodatabase), or a [SQLite Database](https://pro.arcgis.com/en/pro-app/latest/tool-reference/data-management/create-sqlite-database.htm).

##### File Geodatabase

<div class="alert alert-info">
    <b>Note:</b> In the absence of <code>arcpy</code>, the <a href="https://pypi.org/project/Fiona/">Fiona</a> package must be present in your current conda environment to perform this operation.
</div>

In [20]:
# Export to a feature class in File Geodatabase
sdf.spatial.to_featureclass(location="./sedf_data/cities/cities.gdb/major_cities_export")

'C:\\Users\\mohi9282\\Documents\\sedf_guides\\sedf_data\\cities\\cities.gdb\\major_cities_export'

> A Feature Class has been created in a File Geodatabase from the data in the SeDF.

##### Mobile Geodatabase

<div class="alert alert-info">
    <b>Note:</b> This operation can only be performed in an environment that <b>contains</b> <code>arcpy</code>.

</div>

In [21]:
# Export to a feature class in Mobile Geodatabase
sdf.spatial.to_featureclass(location="./sedf_data/cities/cities_mobile.geodatabase/major_cities_export")

'C:\\Users\\mohi9282\\Documents\\sedf_guides\\sedf_data\\cities\\cities_mobile.geodatabase\\main.major_cities_export'

> A Feature Class has been created in a Mobile Geodatabase from the data in the SeDF.

##### SQLite Database

<div class="alert alert-info">
    <b>Note:</b> This operation can only be performed in an environment that <b>contains</b> <code>arcpy</code>.

</div>

In [22]:
# Export to a feature class in SQLite Database
sdf.spatial.to_featureclass(location="./sedf_data/cities/cities.sqlite/major_cities_export")

'C:\\Users\\mohi9282\\Documents\\sedf_guides\\sedf_data\\cities\\cities.sqlite\\major_cities_export'

> A Feature Class has been created in a SQLite Database from the data in the SeDF.

#### Write to a shapefile

The [`to_featureclass()`](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=from_feather#arcgis.features.GeoAccessor.to_featureclass) method can also be used to export spatial data from a _SeDF_ into a shapefile.

<div class="alert alert-info">
    <b>Note:</b> In the absence of <code>arcpy</code>, the <a href="https://pypi.org/project/Fiona/">Fiona</a> package must be present in your current conda environment to perform this operation.
</div>

In [23]:
# Export to a shapefile
sdf.spatial.to_featureclass(location="./sedf_data/cities/major_cities_export.shp")

'C:\\Users\\mohi9282\\Documents\\sedf_guides\\sedf_data\\cities\\major_cities_export.shp'

> A Shapefile has been created from the data in the SeDF.

#### Write to Non-spatial formats
The [`to_table()`](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=from_table#arcgis.features.GeoAccessor.to_table) method can be used to export data from a SeDF into non-spatial formats, such as [csv](https://en.wikipedia.org/wiki/Comma-separated_values) files or [tables](https://pro.arcgis.com/en/pro-app/latest/help/data/geodatabases/overview/table-basics.htm).

##### Write to a csv file

In [24]:
# Export to a csv file
sdf.spatial.to_table(location="./sedf_data/cities/cities_table_export.csv")

'./sedf_data/cities/cities_table_export.csv'

> A csv file has been created from the data in the SeDF.

##### Write to a table in a File Geodatabase

<div class="alert alert-info">
    <b>Note:</b> The operation below can only be performed in an environment that <b>contains</b> <code>arcpy</code>.

</div>

In [25]:
# Export to a table in a File Geodatabase
sdf.spatial.to_table(location="./sedf_data/cities/cities.gdb/cities_table_export")

'C:\\Users\\mohi9282\\Documents\\sedf_guides\\sedf_data\\cities\\cities.gdb\\cities_table_export'

> A table has been created in a File Geodatabase from the data in the SeDF.

## Overwriting GIS Data

The GIS data stored locally can be easily overwritten using the [__Spatially enabled DataFrame__](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html#geoaccessor). Let's take a look.

### Overwrite a Featureclass
The default `overwrite=True` argument in the [`to_featureclass()`](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=from_feather#arcgis.features.GeoAccessor.to_featureclass) method can be used to overwrite an existing feature class from the data in a _SeDF_.

The `major_cities_export` featureclass was created in a section above using `sdf`. We will overwrite this featureclass with a subset of the data from `sdf`.

In [26]:
# Subset the data
sub_df = sdf.iloc[:10,-13:].copy()
sub_df.shape

(10, 13)

In [27]:
# Check head
sub_df.head(2)

NAME  OTHER  OWNER_OCC PLACEFIPS  POP2010  POPULATION  POP_CLASS  \
0      Ammon    307       3205   1601990    13816       15181          6   
1  Blackfoot   1077       2788   1607840    11899       11946          6   

   RENTER_OCC                                              SHAPE  ST STFIPS  \
0        1271  {"x": -12462673.723706163, "y": 5384674.994080...  ID     16   
1        1441  {"x": -12506251.313993266, "y": 5341537.793529...  ID     16   

   VACANT  WHITE  
0     271  13002  
1     318   9893

<div class="alert alert-info">
    <b>Note:</b> In the absence of <code>arcpy</code>, the <a href="https://pypi.org/project/Fiona/">Fiona</a> package must be present in your current conda environment to perform this operation.
</div>

In [28]:
# Export sub_df to the existing major_cities_export featureclass
sub_df.spatial.to_featureclass(location="./sedf_data/cities/cities.gdb/major_cities_export", overwrite=True)

'C:\\Users\\mohi9282\\Documents\\sedf_guides\\sedf_data\\cities\\cities.gdb\\major_cities_export'

In [29]:
# Check if the featureclass is updated
fc_new_df = pd.DataFrame.spatial.from_featureclass(location="./sedf_data/cities/cities.gdb/major_cities_export")
fc_new_df.shape

(10, 14)

> The featureclass has been overwritten with new data.

### Overwrite a table
The default `overwrite=True` argument in the [`to_table()`](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html?highlight=from_table#arcgis.features.GeoAccessor.to_table) method can be used to overwrite an existing non-spatial table from the data in a _SeDF_.

The `cities_table_export` table was created in a section above using `sdf`. We will overwrite this table with a subset of the data `sub_df` defined above.

#### Table in a csv file

In [30]:
# Export sub_df to an existing cities_table_export.csv file
sub_df.spatial.to_table(location="./sedf_data/cities/cities_table_export.csv", overwrite=True)

'./sedf_data/cities/cities_table_export.csv'

In [31]:
# Check if the csv file is updated
tbl_new_df = pd.DataFrame.spatial.from_table(filename="./sedf_data/cities/cities_table_export.csv")
tbl_new_df.shape

(10, 14)

> The csv file has been overwritten with new data.

#### Table in a File Geodatabase

<div class="alert alert-info">
    <b>Note:</b> The operations below can only be performed in an environment that <b>contains</b> <code>arcpy</code>.

</div>

In [32]:
# Export sub_df to an existing table in a File Geodatabase
sub_df.spatial.to_table(location="./sedf_data/cities/cities.gdb/cities_table_export")

'C:\\Users\\mohi9282\\Documents\\sedf_guides\\sedf_data\\cities\\cities.gdb\\cities_table_export'

In [33]:
# Check if the table file is updated
tbl_new_df2 = pd.DataFrame.spatial.from_table(filename="./sedf_data/cities/cities.gdb/cities_table_export")
tbl_new_df2.shape

(10, 13)

> The table file has been overwritten with new data.

## Memory-based Workspace

Writing geoprocessing outputs to memory is an alternative to writing output to a geodatabase or file-based format. It is often significantly faster than writing to on-disk formats. Data written into memory is temporary and is deleted when the application is closed, so it is an ideal location to write intermediate data.

ArcGIS provides two [memory-based workspaces](https://pro.arcgis.com/en/pro-app/latest/help/analysis/geoprocessing/basics/the-in-memory-workspace.htm) where geoprocessing outputs can be written.
- `memory` - is a new memory-based workspace developed for ArcGIS Pro that supports output feature classes, tables, and raster datasets.
- `in_memory` - is the legacy memory-based workspace built for ArcMap that supports output feature classes, tables, and raster datasets.

Let's look at an example of writing to a `memory` workspace. Here, we will:
- write data from SeDF to a `memory` workspace.
- use the data in the memory workspace to generate buffers and export the results to another memory workspace.
- see how results in a memory workspace can be converted to a featureclass.
- delete memory workspaces.

<div class="alert alert-warning">
    <b>Caution:</b> 
    <ul> - Memory-based workspaces do not support geodatabase elements, such as feature datasets, representations, topologies, geometric networks, or network datasets.</ul>
    <ul> - Folders cannot be created in memory-based workspaces.</ul>
    <ul> - Since memory-based workspaces are stored in your system's physical memory, or RAM, your system may run low on memory if you write large datasets into the workspace. This can negatively impact processing performance.</ul>
</div>

<div class="alert alert-info">
    <b>Note:</b> The operations below can only be performed in an environment that <b>contains</b> <code>arcpy</code>.

</div>

In [34]:
# Import arcpy
import arcpy

In [35]:
# Check head
sub_df.head(2)

NAME  OTHER  OWNER_OCC PLACEFIPS  POP2010  POPULATION  POP_CLASS  \
0      Ammon    307       3205   1601990    13816       15181          6   
1  Blackfoot   1077       2788   1607840    11899       11946          6   

   RENTER_OCC                                              SHAPE  ST STFIPS  \
0        1271  {"x": -12462673.723706163, "y": 5384674.994080...  ID     16   
1        1441  {"x": -12506251.313993266, "y": 5341537.793529...  ID     16   

   VACANT  WHITE  
0     271  13002  
1     318   9893

In [36]:
# Write data from SeDF to a memory workspace.
sub_df.spatial.to_featureclass(r"memory\sub_df")

'memory\\sub_df'

In [37]:
# Use data in memory to generate buffers, exporting output to memory 
arcpy.Buffer_analysis(in_features=r"memory\sub_df", 
                      out_feature_class="memory\subBuffers", 
                      buffer_distance_or_field=1)

<Result 'memory\\subBuffers'>

In [38]:
# Read buffer output into a SeDF
buffered_df = pd.DataFrame.spatial.from_featureclass(r"memory\subBuffers")
buffered_df.shape

(10, 16)

In [39]:
# Check head
buffered_df.head(2)

OBJECTID       name  other  owner_occ placefips  pop2010  population  \
0         1      Ammon    307       3205   1601990    13816       15181   
1         2  Blackfoot   1077       2788   1607840    11899       11946   

   pop_class  renter_occ  st stfips  vacant  white  BUFF_DIST  ORIG_FID  \
0          6        1271  ID     16     271  13002        1.0         1   
1          6        1441  ID     16     318   9893        1.0         2   

                                               SHAPE  
0  {"curveRings": [[[-12462673.7237, 5384675.9940...  
1  {"curveRings": [[[-12506251.314, 5341538.79349...

In [43]:
# Convert buffer results to a featureclass
arcpy.Dissolve_management(r"memory\subBuffers", "./sedf_data/cities/cities.gdb/memBuffers2")

<Result '.\\sedf_data\\cities\\cities.gdb\\memBuffers2'>

In [44]:
# Delete the in-memory item
arcpy.Delete_management(r"memory\sub_df")

<Result 'true'>

In [45]:
# Delete the in-memory item
arcpy.Delete_management(r"memory\subBuffers")

<Result 'true'>

## Conclusion

In this guide, we explored how [__Spatially enabled DataFrame__](https://developers.arcgis.com/python/api-reference/arcgis.features.toc.html#geoaccessor) (SeDF) can be used to export spatial data to various formats. We started by exporting the data to web feature layers and to in-memory JSON based formats, such as FeatureSet and FeatureCollection. Next, we explored writing the data to various local data sources, such as a file geodatabase, a mobile geodatabase, an sqlite database, and a shapefile. We also discussed exporting the data to non-spatial formats, such as a csv file or a table. We introduced how the data in local file formats, such as a feature class or a table in a File Geodatabase, can be overwritten using a SeDF. Towards the end, we discussed how the data from SeDF can be exported to in-memory workspaces.

In the next part of the guide series, you will learn about the various properties of a SeDF and how they can be used to pre-process a SeDF.

<div class="alert alert-info">
    <b>Note:</b> Given the importance and popularity of <i>Spatially enabled DataFrame</i>, we are revisiting our documentation for this topic. Our goal is to enhance the existing documentation to showcase the various capabilities of Spatially enabled DataFrame in detail with even more examples this time. 

   Creating quality documentation is time-consuming and exhaustive, but we are committed to providing you with the best experience possible. With that in mind, we will be rolling out the revamped guides on this topic as different parts of a guide series (like the <a href="https://developers.arcgis.com/python/guide/part1-introduction-to-dataengineering/">Data Engineering</a> or <a href="https://developers.arcgis.com/python/guide/part1-introduction-what-is-geometry/">Geometry</a> guide series). This is "part-3" of the guide series for Spatially Enabled DataFrame. You will continue to see the existing documentation as we revamp it to add new parts. Stay tuned for more on this topic.
</div>